{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28518236-23e9-4f0e-9ee3-2532b5823fa2",
   "metadata": {},
   "source": [
    "# 1_Create_MultiModulon_object\n",
    "\n",
    "This notebook demonstrates the first step for multi-species/strain/modality analysis using the MultiModulon package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "845a5330-3189-4ab8-86d8-d69d2ed732f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "from multimodulon import MultiModulon\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Set display options\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "079b426e-33d4-4cc2-a869-c91bea9e862f",
   "metadata": {},
   "source": [
    "## Step 1: Initialize MultiModulon\n",
    "\n",
    "Load data from the Input_Data directory containing expression matrices, gene annotations, and sample metadata for all strains."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "63d6c681-dc25-431a-a050-604c1de91f20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Initializing MultiModulon...\n",
      "\n",
      "Loading from Input_Data:\n",
      "============================================================\n",
      "\n",
      "Nissle_1917:\n",
      "  - Number of genes: 4678\n",
      "  - Number of samples: 183\n",
      "  - Data validation: PASSED\n",
      "\n",
      "O157_H7:\n",
      "  - Number of genes: 5176\n",
      "  - Number of samples: 252\n",
      "  - Data validation: PASSED\n",
      "\n",
      "BL21:\n",
      "  - Number of genes: 4132\n",
      "  - Number of samples: 209\n",
      "  - Data validation: PASSED\n",
      "\n",
      "============================================================\n",
      "Successfully loaded 3 species/strains/modalities\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "# Path to the Input_Data folder\n",
    "input_data_path = './Input_Data'\n",
    "\n",
    "# Initialize MultiModulon object\n",
    "multiModulon = MultiModulon(input_data_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a70867f-52b7-49e5-8acb-f28998191e41",
   "metadata": {},
   "source": [
    "## Step 2: Create Gene Tables\n",
    "\n",
    "Parse GFF files to create gene annotation tables for each strain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5c8b91b4-c883-4f8f-806c-b1d27aadeb00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating gene tables from GFF files...\n",
      "\n",
      "Creating gene tables for all species...\n",
      "============================================================\n",
      "\n",
      "Processing Nissle_1917...\n",
      "  ✓ Created gene table with 4678 genes\n",
      "\n",
      "Processing O157_H7...\n",
      "  ✓ Created gene table with 5137 genes\n",
      "\n",
      "Processing BL21...\n",
      "  ✓ Created gene table with 4132 genes\n",
      "\n",
      "============================================================\n",
      "Gene table creation completed!\n"
     ]
    }
   ],
   "source": [
    "# Create gene tables from GFF files\n",
    "print(\"Creating gene tables from GFF files...\")\n",
    "multiModulon.create_gene_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e251d7f3-97bd-4c74-b1f3-833670678f50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding eggNOG annotations from ./Output_eggnog_mapper\n",
      "============================================================\n",
      "\n",
      "Processing Nissle_1917...\n",
      "  - Reading Nissle_1917.emapper.annotations\n",
      "  ✓ Added eggNOG annotations to 4329/4678 genes\n",
      "\n",
      "Processing O157_H7...\n",
      "  - Reading O157_H7.emapper.annotations\n",
      "  ✓ Added eggNOG annotations to 4681/5137 genes\n",
      "\n",
      "Processing BL21...\n",
      "  - Reading BL21.emapper.annotations\n",
      "  ✓ Added eggNOG annotations to 3993/4132 genes\n",
      "\n",
      "============================================================\n",
      "eggNOG annotation addition completed!\n"
     ]
    }
   ],
   "source": [
    "multiModulon.add_eggnog_annotation(\"./Output_eggnog_mapper\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9705273a-ac80-4138-9bee-653c243d1473",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>ncbi_protein</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>Description</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>EC</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>mobA</th>\n",
       "      <td>AB011548.2</td>\n",
       "      <td>413</td>\n",
       "      <td>736</td>\n",
       "      <td>+</td>\n",
       "      <td>mobA</td>\n",
       "      <td>None</td>\n",
       "      <td>plasmid mobilization</td>\n",
       "      <td>BAA31754.1</td>\n",
       "      <td>1219375.CM002140_gene4213</td>\n",
       "      <td>6.810000e-13</td>\n",
       "      <td>67.4</td>\n",
       "      <td>2E8NX@1|root,33304@2|Bacteria,1NEFP@1224|Prote...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>RHH_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>etpC</th>\n",
       "      <td>AB011549.2</td>\n",
       "      <td>2589</td>\n",
       "      <td>3464</td>\n",
       "      <td>+</td>\n",
       "      <td>etpC</td>\n",
       "      <td>None</td>\n",
       "      <td>Type II secretion pathway related protein</td>\n",
       "      <td>BAA31758.1</td>\n",
       "      <td>155864.EDL933_p0030</td>\n",
       "      <td>7.900000e-186</td>\n",
       "      <td>518.0</td>\n",
       "      <td>COG3031@1|root,COG3031@2|Bacteria,1RD3I@1224|P...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>U</td>\n",
       "      <td>General secretion pathway protein C</td>\n",
       "      <td>gspC</td>\n",
       "      <td>GO:0002790,GO:0006810,GO:0008104,GO:0008150,GO...</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K02452</td>\n",
       "      <td>ko03070,ko05111,map03070,map05111</td>\n",
       "      <td>M00331</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko02044</td>\n",
       "      <td>3.A.15</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>PDZ_2,T2SSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>etpD</th>\n",
       "      <td>AB011549.2</td>\n",
       "      <td>3675</td>\n",
       "      <td>5432</td>\n",
       "      <td>+</td>\n",
       "      <td>etpD</td>\n",
       "      <td>None</td>\n",
       "      <td>Type II secretion pathway related protein</td>\n",
       "      <td>BAA31759.1</td>\n",
       "      <td>155864.EDL933_p0031</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>COG1450@1|root,COG1450@2|Bacteria,1MUUA@1224|P...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>NU</td>\n",
       "      <td>General secretion pathway protein</td>\n",
       "      <td>gspD</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K02453,ko:K03219</td>\n",
       "      <td>ko03070,ko05111,map03070,map05111</td>\n",
       "      <td>M00331,M00332,M00542</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko02044</td>\n",
       "      <td>3.A.15,3.A.6.1,3.A.6.3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Secretin,Secretin_N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>etpE</th>\n",
       "      <td>AB011549.2</td>\n",
       "      <td>5432</td>\n",
       "      <td>6937</td>\n",
       "      <td>+</td>\n",
       "      <td>etpE</td>\n",
       "      <td>None</td>\n",
       "      <td>Type II secretion pathway related protein</td>\n",
       "      <td>BAA31760.1</td>\n",
       "      <td>155864.EDL933_p0032</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>890.0</td>\n",
       "      <td>COG2804@1|root,COG2804@2|Bacteria,1MU7V@1224|P...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>NU</td>\n",
       "      <td>Type II secretory pathway, ATPase PulE Tfp pil...</td>\n",
       "      <td>gspE</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K02454</td>\n",
       "      <td>ko03070,ko05111,map03070,map05111</td>\n",
       "      <td>M00331</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko02044</td>\n",
       "      <td>3.A.15</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>T2SSE,T2SSE_N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>etpF</th>\n",
       "      <td>AB011549.2</td>\n",
       "      <td>6939</td>\n",
       "      <td>8162</td>\n",
       "      <td>+</td>\n",
       "      <td>etpF</td>\n",
       "      <td>None</td>\n",
       "      <td>Type II secretion pathway related protein</td>\n",
       "      <td>BAA31761.1</td>\n",
       "      <td>155864.EDL933_p0033</td>\n",
       "      <td>1.280000e-275</td>\n",
       "      <td>756.0</td>\n",
       "      <td>COG1459@1|root,COG1459@2|Bacteria,1MV4U@1224|P...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>U</td>\n",
       "      <td>General secretion pathway</td>\n",
       "      <td>gspF</td>\n",
       "      <td>GO:0002790,GO:0005575,GO:0005623,GO:0005886,GO...</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K02455,ko:K02505,ko:K02653</td>\n",
       "      <td>ko03070,ko05111,map03070,map05111</td>\n",
       "      <td>M00331</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko02035,ko02044</td>\n",
       "      <td>3.A.15,3.A.15.2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>T2SSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECs_5356</th>\n",
       "      <td>BA000007.3</td>\n",
       "      <td>5492933</td>\n",
       "      <td>5493622</td>\n",
       "      <td>+</td>\n",
       "      <td>creB</td>\n",
       "      <td>None</td>\n",
       "      <td>two-component regulatory system response regul...</td>\n",
       "      <td>BAB38779.1</td>\n",
       "      <td>155864.EDL933_5742</td>\n",
       "      <td>4.320000e-164</td>\n",
       "      <td>459.0</td>\n",
       "      <td>COG0745@1|root,COG0745@2|Bacteria,1MVCB@1224|P...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>K</td>\n",
       "      <td>Transcriptional regulatory protein, C terminal</td>\n",
       "      <td>creB</td>\n",
       "      <td>GO:0000976,GO:0000984,GO:0000986,GO:0000987,GO...</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K07663,ko:K07664</td>\n",
       "      <td>ko02020,map02020</td>\n",
       "      <td>M00449,M00450,M00645,M00646,M00648</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko02022</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Response_reg,Trans_reg_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECs_5357</th>\n",
       "      <td>BA000007.3</td>\n",
       "      <td>5493622</td>\n",
       "      <td>5495046</td>\n",
       "      <td>+</td>\n",
       "      <td>creC</td>\n",
       "      <td>None</td>\n",
       "      <td>two-component system sensor histidine kinase CreC</td>\n",
       "      <td>BAB38780.1</td>\n",
       "      <td>155864.EDL933_5743</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>912.0</td>\n",
       "      <td>COG0642@1|root,COG2205@2|Bacteria,1N17V@1224|P...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>T</td>\n",
       "      <td>Member of the two-component regulatory system ...</td>\n",
       "      <td>creC</td>\n",
       "      <td>GO:0000155,GO:0000160,GO:0003674,GO:0003824,GO...</td>\n",
       "      <td>2.7.13.3</td>\n",
       "      <td>ko:K07641,ko:K07642,ko:K07711,ko:K14980</td>\n",
       "      <td>ko02020,ko02024,map02020,map02024</td>\n",
       "      <td>M00449,M00450,M00502,M00520,M00645,M00646,M00648</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000,ko01001,ko02022</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>HAMP,HATPase_c,HisKA,dCache_3,sCache_3_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECs_5358</th>\n",
       "      <td>BA000007.3</td>\n",
       "      <td>5495104</td>\n",
       "      <td>5496456</td>\n",
       "      <td>+</td>\n",
       "      <td>creD</td>\n",
       "      <td>None</td>\n",
       "      <td>inner membrane protein</td>\n",
       "      <td>BAB38781.1</td>\n",
       "      <td>155864.EDL933_5744</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>877.0</td>\n",
       "      <td>COG4452@1|root,COG4452@2|Bacteria,1MVVR@1224|P...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>V</td>\n",
       "      <td>Inner membrane protein CreD</td>\n",
       "      <td>creD</td>\n",
       "      <td>GO:0005575,GO:0005623,GO:0005886,GO:0016020,GO...</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K06143</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CreD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECs_5359</th>\n",
       "      <td>BA000007.3</td>\n",
       "      <td>5496516</td>\n",
       "      <td>5497232</td>\n",
       "      <td>-</td>\n",
       "      <td>arcA</td>\n",
       "      <td>None</td>\n",
       "      <td>two-component regulatory system response regul...</td>\n",
       "      <td>BAB38782.1</td>\n",
       "      <td>316407.85677140</td>\n",
       "      <td>7.170000e-172</td>\n",
       "      <td>479.0</td>\n",
       "      <td>COG0745@1|root,COG0745@2|Bacteria,1MWJG@1224|P...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>K</td>\n",
       "      <td>It also may be involved in the osmoregulation ...</td>\n",
       "      <td>arcA</td>\n",
       "      <td>GO:0000156,GO:0000160,GO:0000976,GO:0001067,GO...</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K07772,ko:K07773</td>\n",
       "      <td>ko02020,ko02026,map02020,map02026</td>\n",
       "      <td>M00455,M00456</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko02022</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Response_reg,Trans_reg_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECs_5361</th>\n",
       "      <td>BA000007.3</td>\n",
       "      <td>5497868</td>\n",
       "      <td>5498554</td>\n",
       "      <td>+</td>\n",
       "      <td>lasT</td>\n",
       "      <td>None</td>\n",
       "      <td>rRNA methyltransferase</td>\n",
       "      <td>BAB38784.1</td>\n",
       "      <td>155864.EDL933_5746</td>\n",
       "      <td>1.010000e-157</td>\n",
       "      <td>442.0</td>\n",
       "      <td>COG0565@1|root,COG0565@2|Bacteria,1MUKP@1224|P...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>J</td>\n",
       "      <td>Catalyzes the formation of 2'O-methylated cyti...</td>\n",
       "      <td>trmJ</td>\n",
       "      <td>GO:0005575,GO:0005622,GO:0005623,GO:0005737,GO...</td>\n",
       "      <td>2.1.1.200</td>\n",
       "      <td>ko:K02533,ko:K15396</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko01000,ko03016</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SpoU_methylase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5137 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            accession    start      end strand gene_name old_locus_tag  \\\n",
       "locus_tag                                                                \n",
       "mobA       AB011548.2      413      736      +      mobA          None   \n",
       "etpC       AB011549.2     2589     3464      +      etpC          None   \n",
       "etpD       AB011549.2     3675     5432      +      etpD          None   \n",
       "etpE       AB011549.2     5432     6937      +      etpE          None   \n",
       "etpF       AB011549.2     6939     8162      +      etpF          None   \n",
       "...               ...      ...      ...    ...       ...           ...   \n",
       "ECs_5356   BA000007.3  5492933  5493622      +      creB          None   \n",
       "ECs_5357   BA000007.3  5493622  5495046      +      creC          None   \n",
       "ECs_5358   BA000007.3  5495104  5496456      +      creD          None   \n",
       "ECs_5359   BA000007.3  5496516  5497232      -      arcA          None   \n",
       "ECs_5361   BA000007.3  5497868  5498554      +      lasT          None   \n",
       "\n",
       "                                                gene_product ncbi_protein  \\\n",
       "locus_tag                                                                   \n",
       "mobA                                    plasmid mobilization   BAA31754.1   \n",
       "etpC               Type II secretion pathway related protein   BAA31758.1   \n",
       "etpD               Type II secretion pathway related protein   BAA31759.1   \n",
       "etpE               Type II secretion pathway related protein   BAA31760.1   \n",
       "etpF               Type II secretion pathway related protein   BAA31761.1   \n",
       "...                                                      ...          ...   \n",
       "ECs_5356   two-component regulatory system response regul...   BAB38779.1   \n",
       "ECs_5357   two-component system sensor histidine kinase CreC   BAB38780.1   \n",
       "ECs_5358                              inner membrane protein   BAB38781.1   \n",
       "ECs_5359   two-component regulatory system response regul...   BAB38782.1   \n",
       "ECs_5361                              rRNA methyltransferase   BAB38784.1   \n",
       "\n",
       "                       seed_ortholog         evalue   score  \\\n",
       "locus_tag                                                     \n",
       "mobA       1219375.CM002140_gene4213   6.810000e-13    67.4   \n",
       "etpC             155864.EDL933_p0030  7.900000e-186   518.0   \n",
       "etpD             155864.EDL933_p0031   0.000000e+00  1100.0   \n",
       "etpE             155864.EDL933_p0032   0.000000e+00   890.0   \n",
       "etpF             155864.EDL933_p0033  1.280000e-275   756.0   \n",
       "...                              ...            ...     ...   \n",
       "ECs_5356          155864.EDL933_5742  4.320000e-164   459.0   \n",
       "ECs_5357          155864.EDL933_5743   0.000000e+00   912.0   \n",
       "ECs_5358          155864.EDL933_5744   0.000000e+00   877.0   \n",
       "ECs_5359             316407.85677140  7.170000e-172   479.0   \n",
       "ECs_5361          155864.EDL933_5746  1.010000e-157   442.0   \n",
       "\n",
       "                                                  eggNOG_OGs max_annot_lvl  \\\n",
       "locus_tag                                                                    \n",
       "mobA       2E8NX@1|root,33304@2|Bacteria,1NEFP@1224|Prote...    2|Bacteria   \n",
       "etpC       COG3031@1|root,COG3031@2|Bacteria,1RD3I@1224|P...    2|Bacteria   \n",
       "etpD       COG1450@1|root,COG1450@2|Bacteria,1MUUA@1224|P...    2|Bacteria   \n",
       "etpE       COG2804@1|root,COG2804@2|Bacteria,1MU7V@1224|P...    2|Bacteria   \n",
       "etpF       COG1459@1|root,COG1459@2|Bacteria,1MV4U@1224|P...    2|Bacteria   \n",
       "...                                                      ...           ...   \n",
       "ECs_5356   COG0745@1|root,COG0745@2|Bacteria,1MVCB@1224|P...    2|Bacteria   \n",
       "ECs_5357   COG0642@1|root,COG2205@2|Bacteria,1N17V@1224|P...    2|Bacteria   \n",
       "ECs_5358   COG4452@1|root,COG4452@2|Bacteria,1MVVR@1224|P...    2|Bacteria   \n",
       "ECs_5359   COG0745@1|root,COG0745@2|Bacteria,1MWJG@1224|P...    2|Bacteria   \n",
       "ECs_5361   COG0565@1|root,COG0565@2|Bacteria,1MUKP@1224|P...    2|Bacteria   \n",
       "\n",
       "          COG_category                                        Description  \\\n",
       "locus_tag                                                                   \n",
       "mobA                 -                                                  -   \n",
       "etpC                 U                General secretion pathway protein C   \n",
       "etpD                NU                  General secretion pathway protein   \n",
       "etpE                NU  Type II secretory pathway, ATPase PulE Tfp pil...   \n",
       "etpF                 U                          General secretion pathway   \n",
       "...                ...                                                ...   \n",
       "ECs_5356             K     Transcriptional regulatory protein, C terminal   \n",
       "ECs_5357             T  Member of the two-component regulatory system ...   \n",
       "ECs_5358             V                        Inner membrane protein CreD   \n",
       "ECs_5359             K  It also may be involved in the osmoregulation ...   \n",
       "ECs_5361             J  Catalyzes the formation of 2'O-methylated cyti...   \n",
       "\n",
       "          Preferred_name                                                GOs  \\\n",
       "locus_tag                                                                     \n",
       "mobA                   -                                                  -   \n",
       "etpC                gspC  GO:0002790,GO:0006810,GO:0008104,GO:0008150,GO...   \n",
       "etpD                gspD                                                  -   \n",
       "etpE                gspE                                                  -   \n",
       "etpF                gspF  GO:0002790,GO:0005575,GO:0005623,GO:0005886,GO...   \n",
       "...                  ...                                                ...   \n",
       "ECs_5356            creB  GO:0000976,GO:0000984,GO:0000986,GO:0000987,GO...   \n",
       "ECs_5357            creC  GO:0000155,GO:0000160,GO:0003674,GO:0003824,GO...   \n",
       "ECs_5358            creD  GO:0005575,GO:0005623,GO:0005886,GO:0016020,GO...   \n",
       "ECs_5359            arcA  GO:0000156,GO:0000160,GO:0000976,GO:0001067,GO...   \n",
       "ECs_5361            trmJ  GO:0005575,GO:0005622,GO:0005623,GO:0005737,GO...   \n",
       "\n",
       "                  EC                                  KEGG_ko  \\\n",
       "locus_tag                                                       \n",
       "mobA               -                                        -   \n",
       "etpC               -                                ko:K02452   \n",
       "etpD               -                      ko:K02453,ko:K03219   \n",
       "etpE               -                                ko:K02454   \n",
       "etpF               -            ko:K02455,ko:K02505,ko:K02653   \n",
       "...              ...                                      ...   \n",
       "ECs_5356           -                      ko:K07663,ko:K07664   \n",
       "ECs_5357    2.7.13.3  ko:K07641,ko:K07642,ko:K07711,ko:K14980   \n",
       "ECs_5358           -                                ko:K06143   \n",
       "ECs_5359           -                      ko:K07772,ko:K07773   \n",
       "ECs_5361   2.1.1.200                      ko:K02533,ko:K15396   \n",
       "\n",
       "                                KEGG_Pathway  \\\n",
       "locus_tag                                      \n",
       "mobA                                       -   \n",
       "etpC       ko03070,ko05111,map03070,map05111   \n",
       "etpD       ko03070,ko05111,map03070,map05111   \n",
       "etpE       ko03070,ko05111,map03070,map05111   \n",
       "etpF       ko03070,ko05111,map03070,map05111   \n",
       "...                                      ...   \n",
       "ECs_5356                    ko02020,map02020   \n",
       "ECs_5357   ko02020,ko02024,map02020,map02024   \n",
       "ECs_5358                                   -   \n",
       "ECs_5359   ko02020,ko02026,map02020,map02026   \n",
       "ECs_5361                                   -   \n",
       "\n",
       "                                                KEGG_Module KEGG_Reaction  \\\n",
       "locus_tag                                                                   \n",
       "mobA                                                      -             -   \n",
       "etpC                                                 M00331             -   \n",
       "etpD                                   M00331,M00332,M00542             -   \n",
       "etpE                                                 M00331             -   \n",
       "etpF                                                 M00331             -   \n",
       "...                                                     ...           ...   \n",
       "ECs_5356                 M00449,M00450,M00645,M00646,M00648             -   \n",
       "ECs_5357   M00449,M00450,M00502,M00520,M00645,M00646,M00648             -   \n",
       "ECs_5358                                                  -             -   \n",
       "ECs_5359                                      M00455,M00456             -   \n",
       "ECs_5361                                                  -             -   \n",
       "\n",
       "          KEGG_rclass                                            BRITE  \\\n",
       "locus_tag                                                                \n",
       "mobA                -                                                -   \n",
       "etpC                -                  ko00000,ko00001,ko00002,ko02044   \n",
       "etpD                -                  ko00000,ko00001,ko00002,ko02044   \n",
       "etpE                -                  ko00000,ko00001,ko00002,ko02044   \n",
       "etpF                -          ko00000,ko00001,ko00002,ko02035,ko02044   \n",
       "...               ...                                              ...   \n",
       "ECs_5356            -                  ko00000,ko00001,ko00002,ko02022   \n",
       "ECs_5357            -  ko00000,ko00001,ko00002,ko01000,ko01001,ko02022   \n",
       "ECs_5358            -                                          ko00000   \n",
       "ECs_5359            -                  ko00000,ko00001,ko00002,ko02022   \n",
       "ECs_5361            -                          ko00000,ko01000,ko03016   \n",
       "\n",
       "                          KEGG_TC CAZy BiGG_Reaction  \\\n",
       "locus_tag                                              \n",
       "mobA                            -    -             -   \n",
       "etpC                       3.A.15    -             -   \n",
       "etpD       3.A.15,3.A.6.1,3.A.6.3    -             -   \n",
       "etpE                       3.A.15    -             -   \n",
       "etpF              3.A.15,3.A.15.2    -             -   \n",
       "...                           ...  ...           ...   \n",
       "ECs_5356                        -    -             -   \n",
       "ECs_5357                        -    -             -   \n",
       "ECs_5358                        -    -             -   \n",
       "ECs_5359                        -    -             -   \n",
       "ECs_5361                        -    -             -   \n",
       "\n",
       "                                              PFAMs  \n",
       "locus_tag                                            \n",
       "mobA                                          RHH_1  \n",
       "etpC                                    PDZ_2,T2SSC  \n",
       "etpD                            Secretin,Secretin_N  \n",
       "etpE                                  T2SSE,T2SSE_N  \n",
       "etpF                                          T2SSF  \n",
       "...                                             ...  \n",
       "ECs_5356                   Response_reg,Trans_reg_C  \n",
       "ECs_5357   HAMP,HATPase_c,HisKA,dCache_3,sCache_3_2  \n",
       "ECs_5358                                       CreD  \n",
       "ECs_5359                   Response_reg,Trans_reg_C  \n",
       "ECs_5361                             SpoU_methylase  \n",
       "\n",
       "[5137 rows x 28 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiModulon['O157_H7'].gene_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68326d5f-7c1a-4102-bf47-1f6e8c3f20f4",
   "metadata": {},
   "source": [
    "## Step 2: Generate BBH Files\n",
    "\n",
    "Generate Bidirectional Best Hits (BBH) files for ortholog detection between all strain pairs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e13bcf7d-121c-47be-879c-3058fd7e0029",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate BBH files using multiple threads for faster computation\n",
    "output_bbh_path = './Output_BBH'\n",
    "\n",
    "multiModulon.generate_BBH(output_bbh_path, threads=16)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1770e629-845c-4f5c-a6b5-b73d4c684cc9",
   "metadata": {},
   "source": [
    "## Step 3: Align Genes Across Strains\n",
    "\n",
    "Create a unified gene database by aligning genes across all strains using the BBH results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2f743a37-2ed8-4159-9dd5-a2e0c53f9dcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Gene counts in combined_gene_db:\n",
      "  ✓ BL21: 4132 genes (complete)\n",
      "  ✓ Nissle_1917: 4678 genes (complete)\n",
      "  ✓ O157_H7: 5176 genes (complete)\n",
      "\n",
      "✓ All species have complete gene sets in combined_gene_db!\n",
      "\n",
      "Combined gene database shape: (6467, 3)\n",
      "Number of gene groups: 6467\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BL21</th>\n",
       "      <th>Nissle_1917</th>\n",
       "      <th>O157_H7</th>\n",
       "      <th>row_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECD_00001</td>\n",
       "      <td>HW372_19965</td>\n",
       "      <td>None</td>\n",
       "      <td>ECD_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECD_00002</td>\n",
       "      <td>HW372_19960</td>\n",
       "      <td>ECs_0002</td>\n",
       "      <td>ECD_00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECD_00003</td>\n",
       "      <td>HW372_19955</td>\n",
       "      <td>ECs_0003</td>\n",
       "      <td>ECD_00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECD_00004</td>\n",
       "      <td>HW372_19950</td>\n",
       "      <td>ECs_0004</td>\n",
       "      <td>ECD_00004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECD_00005</td>\n",
       "      <td>HW372_19945</td>\n",
       "      <td>ECs_0005</td>\n",
       "      <td>ECD_00005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        BL21  Nissle_1917   O157_H7  row_label\n",
       "0  ECD_00001  HW372_19965      None  ECD_00001\n",
       "1  ECD_00002  HW372_19960  ECs_0002  ECD_00002\n",
       "2  ECD_00003  HW372_19955  ECs_0003  ECD_00003\n",
       "3  ECD_00004  HW372_19950  ECs_0004  ECD_00004\n",
       "4  ECD_00005  HW372_19945  ECs_0005  ECD_00005"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Align genes across all strains\n",
    "output_gene_info_path = './Output_Gene_Info'\n",
    "\n",
    "combined_gene_db = multiModulon.align_genes(\n",
    "    input_bbh_dir=output_bbh_path,\n",
    "    output_dir=output_gene_info_path,\n",
    "    reference_order=['BL21', 'Nissle_1917', 'O157_H7'],  # optional: specify order\n",
    "    # bbh_threshold=90  # optional: minimum percent identity threshold\n",
    ")\n",
    "\n",
    "combined_gene_db.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daff380e-7c0b-43cc-931a-d2dd054f784a",
   "metadata": {},
   "source": [
    "## Step 5: Generate Aligned Expression Matrices\n",
    "\n",
    "Create expression matrices with consistent gene indexing across all strains for multi-view ICA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f0dc005e-3ca9-42e0-8330-35165e2db7cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating aligned expression matrices...\n",
      "\n",
      "Generated aligned X matrices:\n",
      "============================================================\n",
      "BL21: (6467, 209) (4132 non-zero gene groups)\n",
      "Nissle_1917: (6467, 183) (4677 non-zero gene groups)\n",
      "O157_H7: (6467, 252) (5175 non-zero gene groups)\n",
      "============================================================\n",
      "Maximum dimension recommendation: 180\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "# Generate aligned expression matrices\n",
    "print(\"Generating aligned expression matrices...\")\n",
    "multiModulon.generate_X(output_gene_info_path)\n",
    "\n",
    "# The output shows aligned X matrices and dimension recommendations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbf1d608-28b9-471f-bbc2-c60e76b0e77d",
   "metadata": {},
   "source": [
    "## Step 6: Optimize Number of Core Components\n",
    "\n",
    "Use Cohen's d effect size metric to automatically determine the optimal number of core components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8921e6cd-b900-49f2-b06b-1da453733133",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Optimizing number of core components...\n",
      "This will test different values of k and find the optimal number.\n",
      "Optimizing core components for 3 species/strains: ['Nissle_1917', 'O157_H7', 'BL21']\n",
      "Auto-determined max_k = 180 based on minimum samples (183)\n",
      "Using GPU: NVIDIA GeForce RTX 3090\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fbe4fa9d8e4f4f339af6c9da287c2977",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Run 1/1:   0%|          | 0/18 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Optimal k = 90 (robust components passing filter = 31.0)\n",
      "\n",
      "Plot saved to: Output_Optimization_Figures/num_core_optimization.svg\n"
     ]
    }
   ],
   "source": [
    "# Optimize number of core components\n",
    "print(\"Optimizing number of core components...\")\n",
    "print(\"This will test different values of k and find the optimal number.\")\n",
    "\n",
    "optimal_num_core_components = multiModulon.optimize_number_of_core_components(\n",
    "    step=10,                        # Test k = 5, 10, 15, 20, ...\n",
    "    save_path='./Output_Optimization_Figures', # Save plots to directory\n",
    "    fig_size=(7, 5),              # Figure size\n",
    "    num_runs_per_dimension=10,\n",
    "    seed=10\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a1125ae-1d2e-4a90-9fde-ad7d23385d4e",
   "metadata": {},
   "source": [
    "## Step 7: Optimize Number of Unique Components\n",
    "\n",
    "Determine the optimal number of unique (species-specific) components for each strain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "566ea01d-4e66-4ae2-afbe-1975c4b60511",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Optimizing unique components per species...\n",
      "This will test different numbers of unique components for each species.\n",
      "\n",
      "\n",
      "Optimizing unique components with core k = 90\n",
      "\n",
      "============================================================\n",
      "Optimizing unique components for Nissle_1917\n",
      "============================================================\n",
      "Testing a values: [90, 100, 110, 120, 130, 140, 150, 160, 170]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f3c8701a980b40b2bd0cfb9cd382df4e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Testing a values for Nissle_1917:   0%|          | 0/9 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plot saved to: Output_Optimization_Figures/num_unique_Nissle_1917_optimization.svg\n",
      "\n",
      "Optimal a for Nissle_1917: 110 (16 robust unique components)\n",
      "\n",
      "============================================================\n",
      "Optimizing unique components for O157_H7\n",
      "============================================================\n",
      "Testing a values: [90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200, 210, 220, 230, 240]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d2749b1365f6408a92509d7c7c2e8014",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Testing a values for O157_H7:   0%|          | 0/16 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plot saved to: Output_Optimization_Figures/num_unique_O157_H7_optimization.svg\n",
      "\n",
      "Optimal a for O157_H7: 230 (49 robust unique components)\n",
      "\n",
      "============================================================\n",
      "Optimizing unique components for BL21\n",
      "============================================================\n",
      "Testing a values: [90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ca7768c526f449abb016ceaa7c3bff96",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Testing a values for BL21:   0%|          | 0/12 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plot saved to: Output_Optimization_Figures/num_unique_BL21_optimization.svg\n",
      "\n",
      "Optimal a for BL21: 200 (37 robust unique components)\n",
      "\n",
      "============================================================\n",
      "Optimization Summary\n",
      "============================================================\n",
      "Core components (c): 90\n",
      "Nissle_1917: a = 110 (unique components: 20)\n",
      "O157_H7: a = 230 (unique components: 140)\n",
      "BL21: a = 200 (unique components: 110)\n"
     ]
    }
   ],
   "source": [
    "# Optimize unique components for each species\n",
    "print(\"Optimizing unique components per species...\")\n",
    "print(\"This will test different numbers of unique components for each species.\\n\")\n",
    "\n",
    "optimal_unique, optimal_total = multiModulon.optimize_number_of_unique_components(\n",
    "    optimal_num_core_components=optimal_num_core_components,\n",
    "    step=10,\n",
    "    save_path='./Output_Optimization_Figures',\n",
    "    fig_size=(7, 5),\n",
    "    num_runs_per_dimension=10,\n",
    "    seed=10\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5e2f35fd-539b-4042-ad0c-c7162a3cc6dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "90"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "optimal_num_core_components"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1b8422c5-572f-4b5a-8a01-e593cbd68f84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Nissle_1917': 110, 'O157_H7': 230, 'BL21': 200}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "optimal_total"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f0bf2b9-e770-468f-9e6a-1d424dffcabf",
   "metadata": {},
   "source": [
    "## Step 8: Run Robust Multi-view ICA\n",
    "\n",
    "Perform robust multi-view ICA with multiple runs and clustering to identify consistent components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2767e946-712b-44af-a96c-7f87f3832e1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running robust multi-view ICA with clustering...\n",
      "This performs multiple ICA runs and clusters the results for robustness.\n",
      "\n",
      "\n",
      "Running robust multi-view ICA with 10 runs\n",
      "Species: ['Nissle_1917', 'O157_H7', 'BL21']\n",
      "Total components (a): {'Nissle_1917': 110, 'O157_H7': 230, 'BL21': 200}\n",
      "Core components (c): 90\n",
      "\n",
      "Collecting components from 10 runs...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9dd27c74feab4aaab4c1c2e2b6127e3a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "ICA runs:   0%|          | 0/10 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Clustering components...\n",
      "\n",
      "Creating final M matrices with robust components...\n",
      "\n",
      "Saving robust M matrices to species objects...\n",
      "✓ Nissle_1917: (6467, 39) (24 core, 15 unique components)\n",
      "✓ O157_H7: (6467, 63) (24 core, 39 unique components)\n",
      "✓ BL21: (6467, 48) (24 core, 24 unique components)\n",
      "\n",
      "Generating A matrices from robust M matrices...\n",
      "✓ Generated A matrix for Nissle_1917: (39, 183)\n",
      "✓ Generated A matrix for O157_H7: (63, 252)\n",
      "✓ Generated A matrix for BL21: (48, 209)\n",
      "\n",
      "============================================================\n",
      "Robust multi-view ICA completed!\n",
      "============================================================\n",
      "Total core components retained: 24\n",
      "Nissle_1917: 15 unique components\n",
      "O157_H7: 39 unique components\n",
      "BL21: 24 unique components\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "# Run robust multi-view ICA\n",
    "print(\"Running robust multi-view ICA with clustering...\")\n",
    "print(\"This performs multiple ICA runs and clusters the results for robustness.\\n\")\n",
    "\n",
    "M_matrices, A_matrices = multiModulon.run_robust_multiview_ica(\n",
    "    a=optimal_total,                 # Dictionary of total components per species\n",
    "    c=optimal_num_core_components,   # Number of core components\n",
    "    num_runs=10,                     # Number of runs for robustness\n",
    "    seed=100                         # Random seed for reproducibility\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6492ca8-fbc2-43a4-8ddd-9fb225330db0",
   "metadata": {},
   "source": [
    "## Step 9: Optimize thresholds to binarize the M matrices\n",
    "\n",
    "use Otsu's method to calculates thresholds for each component in M matrices across all species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bdec83c8-6514-4247-a5b2-ea8e292f4202",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Optimizing thresholds for Nissle_1917...\n",
      "  Gene mapping: 4678/4678 genes have expression data\n",
      "✓ Optimized thresholds for 39 components\n",
      "  Average genes per component: 25.5\n",
      "\n",
      "Optimizing thresholds for O157_H7...\n",
      "  Gene mapping: 5137/5137 genes have expression data\n",
      "✓ Optimized thresholds for 63 components\n",
      "  Average genes per component: 23.0\n",
      "\n",
      "Optimizing thresholds for BL21...\n",
      "  Gene mapping: 4132/4132 genes have expression data\n",
      "✓ Optimized thresholds for 48 components\n",
      "  Average genes per component: 27.5\n",
      "\n",
      "Threshold optimization completed!\n"
     ]
    }
   ],
   "source": [
    "multiModulon.optimize_M_thresholds(method=\"Otsu's method\", quantile_threshold=95)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8901017-c980-4f9c-ace1-31ea1b8a513f",
   "metadata": {},
   "source": [
    "## Step 10: Save the multiModulon object to json\n",
    "\n",
    "save the multiModulon object to json in the given path and file name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d4cb8cce-d316-4469-b014-2270a907f20d",
   "metadata": {},
   "outputs": [],
   "source": [
    "multiModulon.save_to_json_multimodulon(\"./multiModulon_E_coli_comparison_iMM_2_paper.json.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "eba65f66-d9d8-48f5-ace4-a9081329294c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nissle_1917  :  (6467, 39)\n",
      "O157_H7  :  (6467, 63)\n",
      "BL21  :  (6467, 48)\n"
     ]
    }
   ],
   "source": [
    "for i in multiModulon.species:\n",
    "    print(i, \" : \", multiModulon[i].M.shape)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:multimodulon_dev]",
   "language": "python",
   "name": "conda-env-multimodulon_dev-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
